### PR TITLE
feat(chat): add prompt variables and model chat defaults

### PR DESCRIPTION
**Pull request by rgthelen**
## Summary

Add two new features for improved API usability and reduced repetitive API calls:

### 1. Prompt Variables
Template variables in prompts using `{{variable_name}}` syntax with model-level defaults.

**Model Configuration:**
```yaml
runtime:
  models:
    - name: analyst
      prompts: [persona_prompt]
      prompt_variables:
        persona_name: "DataBot"
        persona_role: "senior data analyst"
        expertise: "financial analysis"
```

**Request Override:**
```json
{
  "messages": [...],
  "model": "analyst",
  "prompt_variables": {
    "persona_name": "CustomBot"
  }
}
```

### 2. Model Chat Defaults
Default chat parameters (temperature, RAG settings, think mode, etc.) at model level.

**Model Configuration:**
```yaml
runtime:
  models:
    - name: analyst
      chat_defaults:
        temperature: 0.3
        rag_enabled: true
        database: "analytics_db"
        think: true
        thinking_budget: 1500
```

## Resolution Priority
- Prompt variables: Request > Model defaults > Empty string
- Chat defaults: Request > Model defaults > None

### New Files
- `server/services/prompt_variable_service.py` - Variable substitution service
- `server/services/chat_defaults_service.py` - Chat defaults resolution service
- `server/tests/test_prompt_variable_service.py` - 18 tests
- `server/tests/test_chat_defaults_service.py` - 10 tests

### Modified Files
- `config/schema.yaml` - Added `prompt_variables` and `chat_defaults` to Model
- `server/api/routers/projects/projects.py` - Integration in chat endpoint
- `server/agents/chat_orchestrator.py` - Apply prompt variables during prompt resolution

## Test Plan
- [x] Unit tests for PromptVariableService (18 tests)
- [x] Unit tests for ChatDefaultsService (10 tests)
- [x] Manual testing with sample llamafarm.yaml

## Backward Compatibility
100% backward compatible - both features are optional with sensible defaults

___

### **PR Type**
Enhancement

___

### **Description**
- Add prompt template variables with {{variable_name}} syntax
  - Model-level defaults via `prompt_variables` in config
  - Request-level overrides in chat requests
  - Priority: Request > Model defaults > Empty string

- Add model-level chat parameter defaults
  - New `chat_defaults` field supporting temperature, RAG, thinking mode, etc.
  - Automatic resolution merging model defaults with request parameters
  - Priority: Request > Model defaults > None

- Integrate both features into chat endpoint and orchestrator

___

### Diagram Walkthrough

```mermaid
flowchart LR
  Config["Model Config<br/>prompt_variables<br/>chat_defaults"]
  Request["Chat Request<br/>prompt_variables<br/>chat params"]
  Resolver["ChatDefaultsService<br/>PromptVariableService"]
  Agent["ChatOrchestratorAgent<br/>with resolved params"]
  Config --> Resolver
  Request --> Resolver
  Resolver --> Agent
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>prompt_variable_service.py</strong><dd><code>Prompt variable substitution service</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

server/services/prompt_variable_service.py

<ul><li>New service for template variable substitution using {{variable_name}} <br>syntax<br> <li> Provides <code>substitute()</code> for single string replacement<br> <li> Provides <code>substitute_messages()</code> for batch message processing with <br>priority resolution<br> <li> Includes utility methods to find variables in content and messages</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/613/files#diff-971ecca98d65bd07e10fbc424707281e2a924cab5f15fae3f780a89b89d747c2">+94/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>chat_defaults_service.py</strong><dd><code>Chat defaults resolution service</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

server/services/chat_defaults_service.py

<ul><li>New service for resolving chat parameter defaults from model config<br> <li> <code>resolve()</code> method merges model defaults with request parameters<br> <li> Supports 12 chat parameters: temperature, top_p, max_tokens, RAG <br>settings, thinking mode, etc.<br> <li> <code>to_request_dict()</code> converts resolved params to dict, filtering None <br>values</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/613/files#diff-de9775cc65b6dcc6ca9b8fd9716087f4a0aaf54d3be220871a387f89a2e2acb8">+98/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>chat_orchestrator.py</strong><dd><code>Integrate prompt variables into orchestrator</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

server/agents/chat_orchestrator.py

<ul><li>Add <code>_prompt_variables</code> field to store request-level prompt variables<br> <li> Accept <code>prompt_variables</code> parameter in <code>__init__()</code> and <code>create_agent()</code><br> <li> Integrate <code>PromptVariableService</code> in <code>_get_prompt_messages_for_model()</code><br> <li> Apply variable substitution with priority: request > model defaults<br> <li> Pass prompt variables through history population</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/613/files#diff-61c784391d5a45c2c6e79b1fd43056b67c8bcbf9d0558f3a8a00e8d4258517a8">+32/-4</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>projects.py</strong><dd><code>Integrate chat defaults and prompt variables in API</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

server/api/routers/projects/projects.py

<ul><li>Add <code>prompt_variables</code> field to <code>ChatRequest</code> model<br> <li> Integrate <code>ChatDefaultsService.resolve()</code> to merge model defaults with <br>request params<br> <li> Pass resolved chat parameters to <code>project_chat_service</code> calls<br> <li> Pass <code>prompt_variables</code> to agent creation in both stateless and stateful <br>modes<br> <li> Import <code>ChatDefaultsService</code> and <code>ModelService</code></ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/613/files#diff-ab4cacdb0344ceca710f889b0d839b684ea3bd780df3f6125661cf30950b287a">+52/-18</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_prompt_variable_service.py</strong><dd><code>Unit tests for prompt variable service</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

server/tests/test_prompt_variable_service.py

<ul><li>18 comprehensive unit tests for <code>PromptVariableService</code><br> <li> Tests for single/multiple variable substitution, missing variables, <br>special characters<br> <li> Tests for message batch processing with priority resolution<br> <li> Tests for variable discovery in content and messages<br> <li> Tests for multiline content and non-string content handling</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/613/files#diff-e9f149e46ee9c81a38c8fe4bc24f323f7df697393e2f2d6a701a7a5dc5eff118">+198/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_chat_defaults_service.py</strong><dd><code>Unit tests for chat defaults service</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

server/tests/test_chat_defaults_service.py

<ul><li>10 comprehensive unit tests for <code>ChatDefaultsService</code><br> <li> Tests for resolution with various combinations of model defaults and <br>request params<br> <li> Tests for priority handling (request overrides model defaults)<br> <li> Tests for <code>to_request_dict()</code> filtering None values while preserving <br>falsy values<br> <li> Tests for all 12 supported chat parameters</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/613/files#diff-4b1bf0ec888555170cf6116dce4823b5406d96b34261b153c0840ed6758fe95d">+206/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>schema.yaml</strong><dd><code>Add prompt variables and chat defaults to schema</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

config/schema.yaml

<ul><li>Add <code>prompt_variables</code> field to Model schema as object with string <br>values<br> <li> Add <code>chat_defaults</code> field to Model schema with 12 optional parameters<br> <li> Document all chat_defaults properties: temperature, top_p, max_tokens, <br>RAG settings, thinking mode, n_ctx<br> <li> Include descriptions and constraints (min/max values) for each <br>parameter</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/613/files#diff-b5e001cf2a392630f0ae6d3499122dd53cb29c671ab0cff2e6e3fbc37bbdce73">+63/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

> [!NOTE]
> Introduces model-level prompt templating and default chat parameters, with full API/runtime integration and docs.
> 
> - Add `variables` (template `{{var}}`) and `chat_defaults` to model schema; document usage across API, configuration, models, and prompts guides
> - Implement `PromptVariableService` for substitution and wire into `ChatOrchestratorAgent` prompt resolution/history
> - Implement `ChatDefaultsService` to merge request params with model `chat_defaults`; use in chat endpoint; plumb `variables` through `ChatRequest` and agent factory
> - Add unit tests for both services covering precedence rules and edge cases
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 12585e297d48b07d3d6fcac0d5bd8ab49acd413e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: config/schema.yaml [hunk 1]
```diff
@@ -200,6 +200,69 @@ properties:
               items:
                 $ref: "#/definitions/Tool"
               default: []
+            prompt_variables:
```
Comment (BobbyRadford) [inline config/schema.yaml]: Why limit scope to prompts by using the `prompt_` prefix. In theory, we could to variable replacement in other parts of the config in the future. I propose just naming this `vars` or `variables`

**Comment by BobbyRadford** [comment]
Closing for lack of activity and PR #672 that was introduced and merged as an alternate approach
